driver/src/main/com/mongodb/reactivestreams/client/gridfs/GridFSUploadPublisher.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2016 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client.gridfs;
+
+import com.mongodb.reactivestreams.client.Success;
+import org.bson.BsonValue;
+import org.bson.types.ObjectId;
+import org.reactivestreams.Publisher;
+
+/**
+ * A GridFS publisher for uploading data into GridFS
+ *
+ * <p>Provides the {@code id} for the file to be uploaded as well as the {@code write} methods of a {@link AsyncOutputStream}</p>
+ *
+ * @param <T> the result type of the publisher
+ * @since 1.13
+ */
+
+public interface GridFSUploadPublisher<T> extends Publisher<T> {
+
+    /**
+     * Gets the {@link ObjectId} for the file to be uploaded
+     *
+     * Throws a {@link com.mongodb.MongoGridFSException} if the file id is not an ObjectId.
+     *
+     * @return the ObjectId for the file to be uploaded
+     */
+    ObjectId getObjectId();
```

Comment:
I'm unclear what this means in the context of a method that returns `GridfsUploadPublisher<ObjectId>`.  Is it the same id?  Can you call this method before subscribing, or before completion of publishing?

Reply:
Yes you can in the same way you could before when using the older API.  The upload publishers should not be reused once completed - as it will cause a duplicate key error.